[WIP] Refactor Istio Metrics to match service->workload transition

douglas-reid opened the pull request:
/hold

This PR is a major overhaul of the default configuration for Mixer-generated Istio telemetry. It was created to address the shift away from `(source|destination).service` attributes to the workload attributes reflecting the newly-available information.

In so doing, it primarily adds the following labels to metrics:
- source_name
- source_namespace
- source_workload
- source_workload_namespace
- source_principal
- source_app
- destination_name
- destination_namespace
- destination_workload
- destination_workload_namespace
- destination_principal
- destination_app
- destination_service_namespace
- request_protocol

This PR also removes the `source_service` label from metrics and alters the `destination_service` expression to use `destination.service.name` attribute.

As part of this cleanup, this PR also renames the default metrics according to the [Prometheus naming conventions](https://prometheus.io/docs/practices/naming/). This means a shift from `istio_request_count` to `istio_requests_total` and `istio_tcp_bytes_sent` to `istio_tcp_sent_bytes_total` for example. The renaming was done in this PR to minimize the number of PRs that affect metrics shape/schema.

A number of accompanying changes must also be made in tandem with metrics changes. They include: servicegraph queries, grafana dashboards, and e2e test updates. Those changes are also included in this PR.

/hold

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6688#" title="Author self-approved">douglas-reid</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)~~ [douglas-reid]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

douglas-reid commented on mixer/adapter/cloudwatch/operatorconfig/cloudwatch.yaml:
```diff
@@ -2,31 +2,29 @@
 apiVersion: "config.istio.io/v1alpha2"
 kind: metric
 metadata:
-  name: requestcount
```
I just copy/pasted from elsewhere. Maybe they had different spacing in different directories for the same content?

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 2f8356fb7334b07cb0a429705f3b801b27ac52f6 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6688/istio-presubmit/13659/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 315a28eaef62f63d6764d496c043f61fad6bcdbc | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6688/istio-presubmit/13669/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@douglas-reid: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | cc063f2295371d6340f3230c187afc79ccb7cc3d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/6688/istio-presubmit/13676/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

bianpengyuan commented on mixer/testdata/config/prometheus.yaml:
```diff
@@ -5,98 +5,156 @@ metadata:
+    - destination_workload_namespace
     - destination_version
----
-apiVersion: "config.istio.io/v1alpha2"
```
Is it intentional to remove these rules? Seems like they are added back anywhere.

jmazzitelli commented:
@douglas-reid is there documentation available that describes these new metrics? Kiali is heavily dependent on these metrics (particularly the request count) and their prom labels...so it is going to need to refactor. having a description of the new metrics and their labels will help.

douglas-reid commented:
@jmazzitelli There will be. If/when this PR is approved/merged, the next step will be to update the docs on the website. Current 1.0 docs: https://preliminary.istio.io/docs/reference/config/policy-and-telemetry/metrics/.

I should note that there is likely to be at least one more PR adding labels to the metrics (to distinguish between client and server side reporting and avoid double-counting).

I expect that this change (or related) will feature heavily in any release notes as well.

gargnupur commented on mixer/testdata/config/accesslog.yaml:
```diff
@@ -16,16 +16,36 @@ spec:
   timestamp: request.time
   variables:
     sourceIp: source.ip | ip("0.0.0.0")
-    destinationService: destination.service | "unknown"
```
why remove destination service?

douglas-reid replied:
`destination.service` attribute is deprecated. This was replaced with `destinationServiceHost` below (as `destination.service.host` is the replacement attribute.

gargnupur replied:
Can we add destination_name: destination.name | "unknown" like it's there in metrics.yaml ?

douglas-reid commented on mixer/adapter/cloudwatch/operatorconfig/cloudwatch.yaml:
```diff
@@ -2,31 +2,29 @@
-  monitored_resource_type: '"UNSPECIFIED"'
+ value: "1"
+ dimensions:
+   source_service: source.service | "unknown"
```
I don't know. I can't really comment on the cloudwatch integration in any detail or with any authority.

douglas-reid commented:
Reviewers: Please take another look. This PR passed all e2e tests last night. I've pushed a few UI tweaks and rebased, but I fully expect everything to pass again.

jeffmendoza commented on install/kubernetes/helm/istio/charts/mixer/templates/config.yaml:
```diff
@@ -234,11 +240,24 @@ metadata:
+    destination_service: destination.service.host | "unknown"
+    destination_service_name: destination.service.name | "unknown"
+    destination_service_namespace: destination.service.namespace | "unknown"
+    request_protocol: api.protocol | context.protocol | "unknown"
```
Looks like the cardinality of the default metrics will be way too high. Any idea on the default Prometheus's configuration capabilities for storage?

douglas-reid replied:
why do you believe that the cardinality is impacted by this change?  large number of deployments?

jeffmendoza replied:
I guess I'm assuming that the # of unique combinations of all the source_* dimensions will be much higher than just the old # of unique source_services, and same for destinatinon_*. Maybe that is not the case. At least we will see increased cardinality based on pod name with the new set, as <src/dest>.name includes the workload instance name, correct?

I'd be nice if we kept the cardinality at: (# of workloads)^2 * (# of response codes)

douglas-reid replied:
`source.service` does not exist any more. so our cardinality has gone up. however, i don't believe that the extra labels increase the cardinality beyond num_workloads too much (name, namespace, owner, and workload should not vary independently).  hopefully, `app` is more-or-less correlated with num_services and is 1:m(workloads).

nothing is currently using the `(source|destination).name` labels. I'd be fine with dropping that, as a way to reduce concerns here (as you could have a deployment that controls a *large* number of pods).

I'd also be fine with dropping `(source|destination).version` as that is highly-variable and somewhat obfuscated by the focus on workload. That just felt a bit radical.

Thoughts?

kyessenov commented on install/kubernetes/helm/istio/charts/mixer/templates/config.yaml:
```diff
@@ -301,10 +362,23 @@ metadata:
     destination_version: destination.labels["version"] | "unknown"
+    destination_service: destination.service.name | "unknown"
+    destination_service_name: destination.service.name | "unknown"
+    destination_service_namespace: destination.service.namespace | "unknown"
```
I'm not sure if we want all of these namespace dimensions. In kubernetes, at least, they are all the same. Doesn't hurt though...

jeffmendoza replied:
Namespace would be good to filter on. People probably want to filter out all the traffic to/from istio-system when looking at their app.

kyessenov replied:
Agree, but we don't need three labels that are always equal.

douglas-reid replied:
@kyessenov in k8s they are the same. i'm not sure that will be the general case necessarily however. I don't have strong feelings here, expect that it seemed somewhat more natural to include the three concepts.

I could see dropping `(source|destination).workload_namespace` at least.

Thoughts?

kyessenov commented on mixer/adapter/cloudwatch/operatorconfig/cloudwatch.yaml:
```diff
@@ -2,31 +2,29 @@
+ dimensions:
+   source_service: source.service | "unknown"
+   source_version: source.labels["version"] | "unknown"
+   destination_service: destination.service | "unknown"
```
change it to destination.service.host while here, or I can do it later...

douglas-reid replied:
If I end up pushing more commits, I'll fix. Good catch.

douglas-reid replied:
done

Stono commented:
We already suffer quite badly due to the number of unique metric combos.

FYI we release 100s of times a day, (440 releases yesterday) so you can
imagine the amount of uniqueness we have.

On Thu, 28 Jun 2018, 9:53 pm Kuat, <notifications@github.com> wrote:

> *@kyessenov* commented on this pull request.
> ------------------------------
>
> In install/kubernetes/helm/istio/charts/mixer/templates/config.yaml
> <https://github.com/istio/istio/pull/6688#discussion_r198982147>:
>
> >      destination_version: destination.labels["version"] | "unknown"
> +    destination_service: destination.service.name | "unknown"
> +    destination_service_name: destination.service.name | "unknown"
> +    destination_service_namespace: destination.service.namespace | "unknown"
>
> I'm not sure if we want all of these namespace dimensions. In kubernetes,
> at least, they are all the same. Doesn't hurt though...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/6688#pullrequestreview-133016148>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABavieAx1RyW2wKS_VPk4fXldSW1sVi2ks5uBUI-gaJpZM4U6kAs>
> .
>

kyessenov commented on mixer/testdata/config/tcpmetrics.yaml:
```diff
@@ -0,0 +1,59 @@
+    destination_principal: destination.principal | "unknown"
+    destination_app: destination.labels["app"] | "unknown"
+    destination_version: destination.labels["version"] | "unknown"
+    destination_service: destination.service.name | "unknown"
```
I think you mean `host` here? Keep in mind some of destination.service.host values are not k8s FQDNs...

douglas-reid replied:
good catch.

douglas-reid commented:
@Stono yeah, 100s of deployments a day is going to be tough to handle with the proposed default configuration.

Do you use the `app` label for your deployments (i can't quite recall now)?  How do you want to segment out deployments?

Is there something in particular that you would recommend here?

douglas-reid commented:
Reviewers: Thanks for all the good feedback, especially since this does involve sifting through a pile of YAML.

I've updated to remove the `*_name` labels from metrics (and add them to logs).

I've also removed the hold on the PR.

PTAL.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio/pull/6688#" title="Author self-approved">douglas-reid</a>*, *<a href="https://github.com/istio/istio/pull/6688#pullrequestreview-133038767" title="LGTM">kyessenov</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/istio/istio/blob/master/OWNERS)~~ [douglas-reid,kyessenov]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

hklai commented:
For some reason the following tests started to fail in postsubmit at this PR.
istio.io/istio/pilot/pkg/serviceregistry/kube.TestServicesChanges
istio.io/istio/broker/pkg/platform/kube/crd.TestBrokerConfigChanges
istio.io/istio/broker/pkg/platform/kube/crd.TestStoreInvariant

See
https://k8s-testgrid.appspot.com/istio-postsubmits#postsubmit
https://prow.istio.io/?type=postsubmit&job=istio-postsubmit

I can clearly see that all presubmit tests passed on this PR. I am not sure why. @douglas-reid could you take a look?

douglas-reid commented:
@hklai those should be completely unrelated to anything that this PR touched. this PR did not change anything `broker` related, nor did it mess with CRDs.  The log messages indicate that the client in those tests had trouble contacting the API server: `client_test.go:73: Post https://35.199.19.58/api/v1/namespaces: dial tcp 35.199.19.58:443: i/o timeout`. I do not have an explanation as to why.

kyessenov commented:
I think the error happens when k8s API server rate limits its clients. The
solution is to move these tests to e2e and treat them as environment
dependent.

On Thu, Jun 28, 2018, 7:27 PM Douglas Reid <notifications@github.com> wrote:

> @hklai <https://github.com/hklai> running those tests locally from the
> same branch that has this PR results in a full pass for each test. the
> presubmit also completes the tests successfully.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/6688#issuecomment-401228710>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AJGIxsgF2AbtMt3AMr1rH4reUj-EtR4Wks5uBZCcgaJpZM4U6kAs>
> .
>

douglas-reid commented:
Local runs:

```
$ git checkout metrics-update
Switched to branch 'metrics-update'
Your branch is up to date with 'origin/metrics-update'.

$ go test ./pilot/pkg/serviceregistry/...
?       istio.io/istio/pilot/pkg/serviceregistry        [no test files]
ok      istio.io/istio/pilot/pkg/serviceregistry/aggregate      0.133s
ok      istio.io/istio/pilot/pkg/serviceregistry/cloudfoundry   5.154s
?       istio.io/istio/pilot/pkg/serviceregistry/cloudfoundry/fakes     [no test files]
ok      istio.io/istio/pilot/pkg/serviceregistry/consul 0.143s
ok      istio.io/istio/pilot/pkg/serviceregistry/eureka 0.263s
ok      istio.io/istio/pilot/pkg/serviceregistry/external       0.173s
ok      istio.io/istio/pilot/pkg/serviceregistry/kube   0.557s
ok      istio.io/istio/pilot/pkg/serviceregistry/memory 0.115s

$ go test ./broker/pkg/platform/kube/...
ok      istio.io/istio/broker/pkg/platform/kube/crd     5.677s
```

These should be unit tests, right? I'm not sure why/how they would be interacting.

As a side note, is anyone doing anything with the broker part of the codebase?
